tracing-tracy/src/config.rs

```diff
@@ -0,0 +1,145 @@
+///     fn formatter(&self) -> &Self::Formatter { &self.formatter }
+///     fn stack_depth(&self) -> u16 { 0 } // Same as the default trait impl.
+///     fn format_fields_in_zone_name(&self) -> bool { true } // Same as the default trait impl.
+/// }
+/// ```
+///
+/// With this sort of setup the compiler will be able to inline calls to `stack_depth` and
+/// `format_fields_in_zone_name` and optimize accordingly.
+pub trait Config {
+    type Formatter: for<'writer> FormatFields<'writer> + 'static;
+
+    /// Use a custom field formatting implementation.
+    fn formatter(&self) -> &Self::Formatter;
+
+    /// Specify the maximum number of stack frames that will be collected.
+    ///
+    /// Note that enabling callstack collection can and will introduce a non-trivial overhead at
+    /// every instrumentation point. Specifying 0 frames (which is the default) will disable stack
+    /// trace collection.
+    ///
+    /// Default implementation returns `0`.
+    fn stack_depth(&self) -> u16 {
+        0
+    }
+
+    /// Specify whether or not to include tracing span fields in the tracy zone name, or to emit
+    /// them as zone text.
+    ///
+    /// The former enables zone analysis along unique span field invocations, while the latter
+    /// aggregates every invocation of a given span into a single zone, irrespective of field
+    /// values.
+    ///
+    /// Default implementation returns `true`.
+    fn format_fields_in_zone_name(&self) -> bool {
+        true
+    }
+
+    /// Apply handling for errors detected by the [`TracyLayer`].
+    ///
+    /// Fundamentally the way the tracing crate and the Tracy profiler work are somewhat
+    /// incompatible in certain ways. For instance, a [`tracing::Span`] can be created on one
+    /// thread and moved to another, where it is cleaned up. Tracy on the other hand expects that
+    /// its eqvivalent concept of zone remains entirely within a thread.
+    ///
+    /// Another example a limitation in `Tracy` where the message length or zone name cannot exceed
+    /// a certain (low) limit of bytes.
+    ///
+    /// Although `tracing_tracy` does it best to paper over these sorts of differences, it can’t
+    /// always make them invisible. In certain cases detecting these sorts of issues is
+    /// straightforward, and it is when `tracing_tracy` will invoke this method to enable users to
+    /// report the issues in whatever way they wish to.
+    ///
+    /// By default a message coloured in red is emitted to the tracy client.
+    fn on_error(&self, client: &Client, error: &'static str) {
+        client.color_message(error, 0xFF000000, self.stack_depth());
+    }
+}
+
+/// A type that implements the [`Config`] trait with runtime-adjustable values.
+///
+/// Ues the [`tracing_subscriber`] [`DefaultFields`] formatter. If not appropriate, consider
+/// implementing the `Config` trait yourself.
+pub struct DynamicConfig {
```

Comment:
Can we also provide a `ConstConfig` that looks like my original PR (so purley generic)? I think that would cover the two extremes. Anything in the middle is going to be pretty application specific, so then it makes sense to require implementing a trait. The benefit of providing `ConstConfig` out of the box is that we can use it and make it the default.

Reply:
I have a feeling that `ConstConfig` is better off defined by the users. While it is possible to define one right now for just `StackDepth` and `FieldsInZoneName` today, this type cannot be expanded in a semver compatible manner as additional configuration options are introduced, leaving it in a spot where:

* it stands out as weirdly incomplete; and
* it isn’t valuable to expand it if there isn’t an upcoming breaking change; and
* any other breaking changes do not force updating of this structure, so it remains incomplete despite the opportunity to bring it up-to-date.

Just to give an example, I have been thinking it might be nice to introduce an ability to add another method to allow computing stack_depth based on `&tracing::Metadata` argument; perhaps even various specializations such as `event_stack_depth`, `span_stack_depth` etc. I’m not yet sure if its a good idea or not, but regardless for dynamic config introducing these in a backward-compatible manner is straightforward; the current methods can pass through configuration to the more fine-grained options to achieve the old effect. For constant config adding, removing or changing a generic in any way that was not coded in the original implementation is a semver break.

To be honest I even pondered if adding a `DynamicConfig` is worth it (over just a `DefaultConfig` that does not provide any way to change its behaviour.) I personally found that whenever a crate provides both options, I eventually end up having my hand forced to migrate to the trait-based configuration anyway.

Reply:
> I have been thinking it might be nice to introduce an ability to add another method to allow computing stack_depth based on &tracing::Metadata argument; perhaps even various specializations such as event_stack_depth, span_stack_depth etc.

Ah, yeah that's a compelling reason to exclude a ConstConfig.

> To be honest I even pondered if adding a DynamicConfig is worth it (over just a DefaultConfig that does not provide any way to change its behaviour.) I personally found that whenever a crate provides both options, I eventually end up having my hand forced to migrate to the trait-based configuration anyway.

I'd be onboard with this. It'll force everybody (including us) to bias towards const values rather than slower defaults. Of course the downside is that the API a little more annoying to use.

Reply:
I pushed up a commit with a removal of `DynamicConfig` (replaced by a `DefaultConfig` that just uses default trait methods for everything.) And made sure to write up an example snippet for people to copy if they need to modify the behaviour in any way.

Reply:
Looks good! And if people complain, then I guess we can add it back.

Reply:
> Just to give an example, I have been thinking it might be nice to introduce an ability to add another method to allow computing `stack_depth` based on `&tracing::Metadata` argument

Any reason to not add the argument on the existing `stack_depth()` method? Having both `fn stack_depth(&self)` and `fn stack_depth_for(&self, _: &tracing::Metadata)` in the API seems suboptimal and kinda confusing.

I'd suggest also adding this argument to `format_fields_in_zone_name`; only having a "global" switch for this can be pretty limiting.

Reply:
Right now perhaps none, since we're making a breaking change anyway, but default methods on a trait like these allow for providing incrementally fancy ways to configure stuff without necessarily exposing every single user to the maximum amount of complexity.

Reply:
In this particular case I'd say the trade-off isn't worth it: there isn't much complexity difference between these two signatures:
```rs
fn stack_depth(&self) -> bool { false }
fn stack_depth(&self, _: &tracing::Metadata<'_>) -> bool { false }
```
And in exchange you now have two very similar methods in the trait, and needs to explain in the documentation which one to implement in which case.